File: 0199-binary-tree-right-side-view/0199-binary-tree-right-side-view.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        solve(root, 0, res);
        return res;
    }

    private void solve(TreeNode root, int level, List<Integer> res) {
        if (root == null) return;

        // If we are visiting this level for the first time, add the node
        if (res.size() == level) {
            res.add(root.val);
        }

        // Visit right subtree first, then left
        solve(root.right, level + 1, res);
        solve(root.left, level + 1, res);
    }
}

File: 0199-binary-tree-right-side-view/0199-binary-tree-right-side-view.cpp
/**
 *Definition for a binary tree node.
 *struct TreeNode {
 *    int val;
 *    TreeNode * left;
 *    TreeNode * right;
 *    TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 *};
 */
class Solution
{
    private:
        void solve(TreeNode *root, int level, vector<int> &res)
        {
            if (!root) return;
           	// At each level we see only the rightmost node of that respective level. if no right node exists then go for the left node. At each level we put only one the one element into the vector res
            if (res.size() == level) res.push_back(root->val);
            solve(root->right, level + 1, res);
            solve(root->left, level + 1, res);
        }
    public:
        vector<int> rightSideView(TreeNode *root)
        {
            if (!root) return {};
            vector<int> res;
            solve(root, 0, res);
            return res;
        }
};